### PR TITLE
Report accidents only if you are an eyewitness

### LINKED ISSUE #52832: Character reports seeing vehicle collisions outside of one's field of view
### Describe the bug

`The Wreckage's frame collides with Wreckage's V6 engine.`

### Steps To Reproduce

Load the attached save file and `.`wait a minute.

### Expected behavior

No strange collisions when i'm chilling in a toilet.

### Screenshots

screenshot-toilet-collision

### Versions and configuration

- OS: Linux
    - OS Version: <unknown>
- Game Version: 0.F-3857-g5eb505ae9a-dirty [64-bit]
- Graphics Version: Curses
- Game Language: English [en]
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    No Fungal Growth [no_fungal_growth],
    Bionic Professions [package_bionic_professions]
]

### Additional context

[2021.11.10.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/7536337/2021.11.10.zip)

### PR DESCRIPTION
**Pull request by wapcaplet**
#### Summary

Bugfixes "Report accidents only if you are an eyewitness"

#### Purpose of change

Fixes #52832

#### Describe the solution

When two vehicle parts collide, ensure the avatar can actually see them before writing log messages about them.

Also prevent a possible division by zero, and instead print a debug message "The vehicle collides with bugs in the code" if there are no parts to collide with. I'm not sure how this part passed clang-tidy checks before but it was getting red-flagged for whatever reason.

#### Describe alternatives you've considered

I may clean up a few other `add_msg` cases in here while I'm thinking about it.

#### Testing

- Spawn a stationary vehicle to be the target
- Spawn a vehicle you can drive
- Hop in and accelerate toward the parked vehicle
- Teleport away before collision
- Wear a blindfold while the collision unfolds - hear sounds, but don't see colliding parts

### CODE REVIEW

File: src/map.cpp [hunk 1]
```diff
@@ -732,12 +732,22 @@ float map::vehicle_vehicle_collision( vehicle &veh, vehicle &veh2,
...
+    // Check whether avatar sees the collision, and log a message if so
+    const avatar &you = get_avatar();
+    const tripoint part1_pos = veh.global_part_pos3( c.part );
+    const tripoint part2_pos = veh.global_part_pos3( c.target_part );
+    if( you.sees( part1_pos ) || you.sees( part2_pos ) ) {
+        //~ %1$s: first vehicle name (without "the")
+        //~ %2$s: first part name
+        //~ %3$s: second vehicle display name (with "the")
+        //~ %4$s: second part name
+        add_msg( m_bad, _( "The %1$s's %2$s collides with %3$s's %4$s." ),
```
Comment (wapcaplet) [inline src/map.cpp:746]: Is it OK to split up translation notes across multiple lines? This is hard to cram into a single comment.

Comment (BrettDong) [reply src/map.cpp:746]: Yes, all four lines are recognized.

**Comment by PatrikLundell** [comment]
There's an old crash bug involving collisions that will be even harder to work around when you don't get these collision messages. As far as I can tell, sufficiently large zombies can push vehicles away to collide with things, possibly non vehicles (such as the gate of a private resort), resulting in a crash (some of the time). However, your division by zero detection might be what would avoid this crash.

**Comment by PatrikLundell** [comment]
I don't think I've reported it. At the time I'd submitted several reports that hadn't seen any indication whatsoever of processing apart from labels, so it felt pointless to just throw another one into the void. It doesn't help that save games quickly grow to a size that's too large to be attached to the report even with maximum compression.

What's happened in those cases is that I get a report that a vehicle crashes into something and then the game crashes. As these vehicles have been out of sight means that I haven't been able to see what actually happens, but my conclusion is that vehicles are shoved away by huge zombies to crash into something that somehow cause a collision with properties that aren't handled. In the private resort case, I believe the thing crashed into was a gate (which has the property of being possible to be raised, which might be a factor), and the vehicle was a limo (probably not relevant). In another case I have no idea which vehicle was pushed where, as it was on some street at the other side of the block. In both cases I managed to work around the problem by avoiding coming from a direction that might, somehow, cause the (unseen) zombies to not move in the direction of the (unseen) vehicle. In the private resort case it meant not driving in front of the resort and enter through the breach to the side made by zombies and then kill off the big ones (or, rather, all of them) from that direction, rather than having them drawn into the yard.

**Comment by wapcaplet** [comment]
Sorry, I don't know what to suggest except to open an issue reporting that behavior. I realize we have a lot of open issues, but unreported issues are the least likely to be addressed.

**Comment by PatrikLundell** [comment]
I will if I encounter it again. I don't think a vague report from 0.E2 is of much use, in particular if the division by zero check would prevent a crash.

**Pull request merged**